README.md
<!--
 * @Author: Chengbotao
 * @Date: 2022-06-13 23:52:33
-->

![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/chengbotao/billows)
![GitHub](https://img.shields.io/github/license/chengbotao/billows)
![GitHub Repo stars](https://img.shields.io/github/stars/chengbotao/billows)

# Billows

**常用的工具函数集合**

## Install

```bash
npm install billows
```

## Usage

```ts
// 全量引入
import * as billows from 'billows';
billows.typeOf('chengbotao'); // "string"

// 差量使用
import { typeOf } from 'billows';
typeOf('chengbotao'); // "string"
```

| 函数名                                                                                            | 说明                                                            |
| ------------------------------------------------------------------------------------------------- | --------------------------------------------------------------- |
| [`typeOf`](https://chengbotao.github.io/chengbotao/billows/typeOf.html)                           | 返回变量类型                                                    |
| [`instanceOf`](https://chengbotao.github.io/chengbotao/billows/instanceOf.html)                   | 检测构造函数的 `prototype` 属性是否出现在某个实例对象的原型链上 |
| [`setValueByReference`](https://chengbotao.github.io/chengbotao/billows/setValueByReference.html) | 通过引用设置值的函数                                            |
| [`getValueByReference`](https://chengbotao.github.io/chengbotao/billows/getValueByReference.html) | 通过引用获取值的函数                                            |
| [`deepClone`](https://chengbotao.github.io/chengbotao/billows/deepClone.html)                     | 深拷贝函数                                                      |
| [`treeToFlat`](https://chengbotao.github.io/chengbotao/billows/treeToFlat.html)                   | 树结构转扁平化结构                                              |
| [`partial`](https://chengbotao.github.io/chengbotao/billows/partial.html)                         | 偏函数                                                          |
| [`getNodeFromTree`](https://chengbotao.github.io/chengbotao/billows/getNodeFromTree.html)         | 根据唯一标志从树结构中获取节点                                  |
| [`flatToTree`](https://chengbotao.github.io/chengbotao/billows/flatToTree.html)                   | 扁平化转结构转树结构                                            |
| [`curry`](https://chengbotao.github.io/chengbotao/billows/curry.html)                             | 函数柯里化                                                      |
| [`buildURL`](https://chengbotao.github.io/chengbotao/billows/buildURL.html)                       | 由 `url` 和 `params` 生成 URL                                   |
| [`debounce`](https://chengbotao.github.io/chengbotao/billows/debounce.html)                       | 函数防抖                                                        |
| [`encodeURIC`](https://chengbotao.github.io/chengbotao/billows/encodeURIC.html) | 有过滤特殊字符功能的 `encodeURIComponent`                       |
| [`isURLSearchParams`](https://chengbotao.github.io/chengbotao/billows/isURLSearchParams.html)     | 判断是否是 `URLSearchParams`                                    |
| [`jsonp`](https://chengbotao.github.io/chengbotao/billows/jsonp.html)                             | `promise` 封装 `jsonp`                                          |
| [`throttle`](https://chengbotao.github.io/chengbotao/billows/throttle.html)                       | 函数节流                                                        |
| [`parseUrl`](https://chengbotao.github.io/chengbotao/billows/parseurl.html)                       | 解析 `URL` 参数                                                  |

## Docs

[文档地址](https://chengbotao.github.io/chengbotao/billows)

## License

**MIT**


CHANGELOG.md
<!--
 * @Author: Chengbotao
 * @Date: 2022-06-15 09:27:24
-->

## 1.0.4 (2022-07-11)


### Features

- **parseUrl:** 解析 URL 参数 ([1cb5a20](https://github.com/chengbotao/billows/commit/1cb5a20f4db4fc38df7e9dab61b0151fcab84454))

# 1.0.3 (2022-06-24)

### Features

- **buildURL:** 由 `url` 和 `params` 生成 URL ([bbeda5c](https://github.com/chengbotao/billows/commit/bbeda5cf111c0bf2ad41071178081f6ecbb2aca9))
- **debounce:** 函数防抖 ([7bab330](https://github.com/chengbotao/billows/commit/7bab330422433e78176013b11660f032302f6ea9))
- **encodeURIC:** 有过滤特殊字符功能的 `encodeURIComponent` ([723c04a](https://github.com/chengbotao/billows/commit/723c04af531a40297b1f13db0e2a7e6b2402ec5a))
- **isURLSearchParams:** 判断是否是 `URLSearchParams` ([75e7e34](https://github.com/chengbotao/billows/commit/75e7e344eabac10996fd4a78a2f611e6d55ec1a5))
- **jsonp:** `promise` 封装 `jsonp` ([9d5a2e0](https://github.com/chengbotao/billows/commit/9d5a2e018214bba7fd4d40bf00ad34acdbefe23e))
- **throttle:** 函数节流 ([01d8fc7](https://github.com/chengbotao/billows/commit/01d8fc75ce0db8b7c1fc20375928e5b05b5f6820))

# 1.0.2 (2022-06-18)

### Refactor

- **deepClone**: 深克隆函数,对内置对象的处理 ([3c300a2](https://github.com/chengbotao/billows/commit/3c300a2))
- **getNodeFromTree**: 重构获取树结构中节点函数 ([f049b7c](https://github.com/chengbotao/billows/commit/f049b7c))
- **instanceOf**: 更新 instanceof 函数参数命名 ([b1ca43b](https://github.com/chengbotao/billows/commit/b1ca43b))
- **treeToFlat**: 树结构转扁平化结构代码重构 ([8694df2](https://github.com/chengbotao/billows/commit/8694df2))
- **treeToFlat**: 重构树结构转扁平化结构 ([d53c498](https://github.com/chengbotao/billows/commit/d53c498))
- **typeOf**: 修改参数命名 ([c9bce64](https://github.com/chengbotao/billows/commit/c9bce64))

# 1.0.1 (2022-06-14)

### Features

- **curry:** 函数柯里化 ([df88a1a](https://github.com/chengbotao/billows/commit/df88a1afef0a8e426a12767754f348ab03dd336a))
- **flatToTree:** 扁平化转结构转树结构 ([5be1ccc](https://github.com/chengbotao/billows/commit/5be1ccc5328b1d0e5c143d82dbfcb7c085d30e55))
- **getNodeFromTree:** 根据唯一标志从树结构中获取节点 ([08f1552](https://github.com/chengbotao/billows/commit/08f1552ab1e7a888dbb0761ad4398868894f6d47))
- **partial:** 偏函数 ([ab8fa1f](https://github.com/chengbotao/billows/commit/ab8fa1fe253b7bd7e96d5c5a2f88240ea2d05eab))
- **treeToFlat:** 树结构转扁平化结构 ([0ecdb1b](https://github.com/chengbotao/billows/commit/0ecdb1b56f677e6f3d29703ddd45dfc3aca53097))

# 1.0.0 (2022-06-13)

### Features

- **deepClone:** 深拷贝函数 ([232ff63](https://github.com/chengbotao/billows/commit/232ff63435dc6b16707d43fd5f978aff40765051))
- **getValueByReference:** 通过引用获取值的函数 ([55fe9e6](https://github.com/chengbotao/billows/commit/55fe9e65b5dd932ac5906e8fae84e2ef8104f276))
- **instanceOf:** instanceof 函数 ([18c1674](https://github.com/chengbotao/billows/commit/18c1674a67935029f04d29552cb92c5a3e0d8007))
- **setValueByReference:** 通过引用设置值的函数 ([ccff921](https://github.com/chengbotao/billows/commit/ccff921713d956442822da1a165d0301daccf1a3))
- **typeOf:** 类型判断函数 ([e653076](https://github.com/chengbotao/billows/commit/e653076a1c5d12a8092f2b8db8f9755ce04addd2))


test/typeof.spec.ts
/*
 * @Author: Chengbotao
 * @Date: 2022-06-08 17:04:44
 */

/* eslint-disable no-undef */
import { typeOf } from '../packages/typeof';

describe('类型(typeOf)', () => {
  test('typeOf("chengbotao") => "string"', () => {
    expect(typeOf('chengbotao')).toBe('string');
  });
  test('typeOf(28) => "number"', () => {
    expect(typeOf(28)).toBe('number');
  });
  test('typeOf(true) => "boolean"', () => {
    expect(typeOf(true)).toBe('boolean');
    expect(typeOf(false)).toBe('boolean');
  });
  test('typeOf(undefined) => "undefined"', () => {
    expect(typeOf(undefined)).toBe('undefined');
  });
  test('typeOf(null) => "null"', () => {
    expect(typeOf(null)).toBe('null');
  });
  test('typeOf(new Date()) => "date"', () => {
    expect(typeOf(new Date())).toBe('date');
  });
  test('typeOf([]) => "array"', () => {
    expect(typeOf([])).toBe('array');
  });
  test('typeOf({}) => "object"', () => {
    expect(typeOf({})).toBe('object');
  });
  test('typeOf(function fun() {}) => "function', () => {
    function fun() {}
    expect(typeOf(fun)).toBe('function');
  });
  test('typeOf(Promise.resolve()) => "promise"', () => {
    expect(typeOf(Promise.resolve())).toBe('promise');
  });
});


test/deepclone.spec.ts
/*
 * @Author: Chengbotao
 * @Date: 2022-06-10 14:49:25
 */
/* eslint-disable no-undef */
import { deepClone } from '../packages/deepclone';

describe('深克隆(deepClone)', () => {
  const obj = {
    name: 'chengbotao',
    age: 28,
  };

  test('deepClone(honda, Car) => true', () => {
    expect(deepClone(obj)).toEqual({ name: 'chengbotao', age: 28 });
  });
});


test/instanceof.spec.ts
/*
 * @Author: Chengbotao
 * @Date: 2022-06-10 14:25:20
 */
/* eslint-disable no-undef */
import { instanceOf } from '../packages/instanceof';

describe('实例(instanceOf)', () => {
  class Car {
    public make: any;
    public model: any;
    public year: any;
    constructor(make: any, model: any, year: any) {
      this.make = make;
    }
  }
  const honda = new Car('Honda', 'Accord', 1998);

  test('instanceOf(honda, Car) => true', () => {
    expect(instanceOf(honda, Car)).toBeTruthy();
  });
});


.cz-config.js
/*
 * @Author: Chengbotao
 * @Date: 2022-06-08 01:08:44
 */
module.exports = {
  types: [
    { value: "feat", name: "✨ Features | 新功能" },
    { value: "fix", name: "🐛 Bug Fixes | Bug 修复" },
    { value: "init", name: "🎉 Init | 初始化" },
    { value: "docs", name: "✏️ Documentation | 文档" },
    { value: "style", name: "💄 Styles | 风格" },
    { value: "refactor", name: "♻️ Code Refactoring | 代码重构" },
    { value: "perf", name: "⚡ Performance Improvements | 性能优化" },
    { value: "test", name: "✅ Tests | 测试" },
    { value: "revert", name: "⏪ Revert | 回退", hidden: true },
    { value: "build", name: "📦‍ Build System | 打包构建" },
    { value: "chore", name: "🚀 Chore | 构建/工程依赖/工具" },
    { value: "ci", name: "👷 Continuous Integration | CI 配置" }
  ],

  // scopes: [{ name: 'accounts' }, { name: 'admin' }, { name: 'exampleScope' }, { name: 'changeMe' }],

  allowTicketNumber: false,
  isTicketNumberRequired: false,
  ticketNumberPrefix: 'TICKET-',
  ticketNumberRegExp: '\\d{1,5}',

  // it needs to match the value for field type. Eg.: 'fix'
  /*
  scopeOverrides: {
    fix: [
      {name: 'merge'},
      {name: 'style'},
      {name: 'e2eTest'},
      {name: 'unitTest'}
    ]
  },
  */
  // override the messages, defaults are as follows
  messages: {
    type: "选择要提交的更改类型:",
    scope: '\n表示此更改的范围(可选):',
    // used if allowCustomScopes is true
    customScope: '表示此更改的范围(自定义):',
    subject: '写一个简短的，命令式的时态描述变化:\n',
    body: '提供更改的较长描述(可选)。使用"|"换行:\n',
    breaking: '列出任何重大更改(可选):\n',
    footer: '列出此更改关闭的任何问题(可选). E.g.: #31, #34:\n',
    confirmCommit: '您确定要继续上面的提交吗?',
  },

  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix'],
  // skip any questions you want
  skipQuestions: ['body'],

  // limit subject length
  subjectLimit: 100,
  // breaklineChar: '|', // It is supported for fields body and footer.
  // footerPrefix : 'ISSUES CLOSED:'
  // askForBreakingChangeFirst : true, // default is false
};

jest.config.ts
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "C:\\Users\\chengbotao\\AppData\\Local\\Temp\\jest",

  // Automatically clear mock calls, instances, contexts and results before every test
  // clearMocks: false,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  coverageDirectory: 'coverage',

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "\\\\node_modules\\\\"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  // coverageProvider: "babel",

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // The default configuration for fake timers
  // fakeTimers: {
  //   "enableGlobally": false
  // },

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "mjs",
  //   "cjs",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "json",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest',

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state before every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state and implementation before every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: 'jest-environment-node',

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "\\\\node_modules\\\\"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jest-circus/runner",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "\\\\node_modules\\\\",
  //   "\\.pnp\\.[^\\\\]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


rollup.config.ts
/*
 * @Author: Chengbotao
 * @Date: 2022-06-05 18:00:17
 */
import { resolve } from 'path';
import pkg from './package.json';

import json from '@rollup/plugin-json';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import alias from '@rollup/plugin-alias';
import eslint from '@rollup/plugin-eslint';
import typescript from '@rollup/plugin-typescript';
import { babel } from '@rollup/plugin-babel';

const { version } = pkg;
const outputConf = {
    banner: `/* Billows version ${version} */`,
    footer: `/* Follow me on GitHub! @chengbotao */`,
};

export default {
    input: 'packages/index.ts',
    output: [
        {
            file: resolve(__dirname, 'dist', 'billows.esm.js'),
            format: 'esm',
            ...outputConf,
        },
        {
            file: resolve(__dirname, 'dist', 'billows.js'),
            format: 'umd',
            name: 'billows',
            ...outputConf,
        },
    ],
    plugins: [
        alias({
            entries: {
                packages: resolve(__dirname, 'packages'),
            },
        }),
        nodeResolve({
            extensions: ['.ts', '.tsx', '.mjs', '.js', '.jsx', '.json'],
        }),
        commonjs(),
        eslint({
            // fix: true,
            include: ['packages/**/*.ts'],
            exclude: ['node_modules/**', 'dist', 'test'],
        }),
        babel({
            babelHelpers: 'runtime',
            exclude: 'node_modules/**',
            extensions: ['.js', '.jsx', '.es6', '.es', '.mjs', '.ts', 'tsx'],
        }),
        typescript(),
        json(),
    ],
};
